sdk/core/System.ClientModel/src/Auth/TokenProvider.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel;
+
+/// <summary>
+/// Represents a provider that can provide a token.
+/// </summary>
+public abstract class TokenProvider<TContext> where TContext : ITokenContext
```

Comment:
Example of TokenProvider generic type that implements various interfaces to indicate the capabilities it supports.